doc/dev/kdb-operations.md

```diff
@@ -0,0 +1,159 @@
+# Proposal: KDB Operations
+
+There are four main _operations_ in `libelektra-kdb`: `open`, `get`, `set` and `close`.
+For each of these there is a `kdb*` function the user calls to trigger the operation and plugins implement a `elektra<Plugin>*` function for each of the operations they support (at least `get`).
+
+Additionally, plugins may implement `elektra<Plugin>Commit` and `elektra<Plugin>Error`.
+These are part of the `set` operation and there is no corresponding `kdbCommit` or `kdbError` function available in `libelektra-kdb`.
+
+The operations `get` and `set` also have different _phases_:
+
+- The `get` operation has: `init`, `resolver`, `cachecheck`, `prestorage`, `storage` and `poststorage`.
+- The `set` operation has: `resolver`, `prestorage`, `storage` and `poststorage` followed by `precommit`, `commit` and `postcommit` if the previous phases where successful or by `prerollback`, `rollback` and `postrollback` if the previous phases failed.
+
+These phases are implemented by a backend plugin.
+Read the [Documentation on Backend Plugins](./backend-plugins.md) for more information on what backend plugins do.
+
+## `open` Operation
+
+The `open` operation implemented in `kdbOpen` is the first thing that happens to all `KDB` instances.
+
+The basic flow of this operation is:
+
+1. Create empty `KDB` instance
+2. Configure `KDB` instance for bootstrap
+3. Run bootstrap `get` operation
+4. Mount global plugins
+5. Process contract
+6. Parse mountpoints
+7. Configure `KDB` instance with real mountpoints
+8. Add hardcoded mountpoints to `KDB` instance
+
+Namespaces in mountpoint configs:
+
+- cascading mountpoints act as an alias for identical `proc:/`, `dir:/`, `user:/`, `system:/` and `default:/` mountpoints.
+- `meta:/` mountpoints are illegal
+- explicit `default:/` mountpoints are illegal
+  They can only exist via cascading mountpoints.
+  This is so that default values can be checked and validated by plugins.
+- `spec:/` mountpoints may exist, but they are treated differently during `get` and `set`
+- `proc:/` mountpoints may exist, but they are read-only (see below)
+- `dir:/`, `user:/` and `system:/` mountpoints can be created without restrictions
+
+Other restrictions:
+
+- Creating mountpoints for root keys (`system:/`, `user:/`, etc.) is not allowed.
```

Comment:
I don't see why this would be useful. Maybe to change the default storage format, but could be done via a different mechanism.

My main concern here was that to me it doesn't make sense to mount anything more than the bare minimum (i.e. a simple backend with resolver and storage) to the root keys. If you have some config that needs more plugins, just create a named mountpoint for that.

However, if you really want mounting to root keys to work, I'm sure we can make it work. I seem to remember that I added this restriction because it made some piece of code easier to write and I didn't think it would matter.

Reply:
Not only the default storage format (this could be done by changing symlinks) but to add any plugin additional globally. If you have only one mountpoint, it is probably /.

Reply:
But why? What is the _specific use-case_ for adding a plugin to a root mountpoint?

Note: Adding a plugin to the root mountpoint is not what I would understand as adding a plugin globally. The root mountpoint is only used as the fallback, if there isn't a more specific mountpoint to store a key. Globally to me implies that it applies to every mountpoint, not just the fallback ones.

Reply:
The specific use case is, for example, that you have an embedded system with several applications and you want to mount a few (non-global) plugins for all the applications.

Reply:
I'm assuming each application has it's own separate mountpoint, e.g. `/sw/lcdproc/lcdd`, `/sw/lcdproc/lcdproc`. This is how Elektra is intended to be used and that is what we should support.

In this case, adding a plugin to `system:/` wouldn't do anything since neither LCDd nor lcdproc, would ever write a key to that mountpoint. What you should do instead is add the plugins to the specific mountpoints for the applications.

What you actually need to simplify this use-case is a way to define a set of default plugins, that are added to every mountpoint. This is not currently possible. It would be a totally new feature.

---

Maybe you are talking about a different case, where no mountpoints would be created for the various applications and they all use the default root mountpoints. This is not how Elektra is designed to be used. Different applications have different needs and therefore should use different mountpoints.

If you want to combine all your applications into a kind of embedded system image that has a fixed set of requirements, you could simply create a mountpoint that encompasses all your applications, e.g. `/sw/lcdproc`. This mountpoint could then be configured.

Reply:
You are describing a different use case (desktop). In embedded systems you might not want a mountpoint for every application as there are many applications with little configs (e.g. only "enable").

Reply:
Once again: Elektra tries to be super flexible and serve everyone. This is still the biggest gripe I have with Elektra. The target audience is just: Everyone. This is simply not feasible for 1.0 with such a small time. A huge and long-standing project like GCC or glibc, might be able to serve both desktop and embedded systems well (*), even though the too worlds are completely different.

---

(*) Even in this case there are compilers and libcs specialised on embedded systems, because the are just so incredibly different.

---

Anyway, you have not addressed, why I couldn't just create a single mountpoint `/foo` for the whole embedded system. Plus maybe a few extra mountpoints for bigger applications with larger configs.

Also, an embedded system probably compiles Elektra from source anyway, so if the default backend (including all plugins not just storage+resolver) can be changed at compile-time this should also be enough for embedded systems.

Reply:
Discussion result: We will allow mounting to root names. Mounting to `/elektra` (in any namespace) will still be illegal. `/elektra` will always use a hardcoded config.